### PR TITLE
Redis refactor

### PR DESCRIPTION
**Pull request by iriberri**
#273

### CODE REVIEW

**Comment by rafatower** [comment]
What's here is a big design change:
- move down the stack all low level concerns: "db connections" and redis connections
- avoid too much global state scattered through `GD` var
- encapsulate the redis connection creation in a factory

The whole thing so far is working E2E for mapzen geocoder. We'd like you to take a look before it goes any further, cause this refactor affects the whole server, before arch details are hidden in a pile of commits.

File: server/lib/python/cartodb_services/cartodb_services/__init__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#     import cartodb_services
+#     cartodb_services.init(GD)
+#
+#     # rest of the code here
+#
+# $$ LANGUAGE plpythonu;
+
+def init(GD):
+    import plpy
+    plpy.GD = GD
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/__init__.py]: IMHO GD should be part of `plpy` module and this is a workaround for that design flaw in plpythonu language.

File: server/lib/python/cartodb_services/cartodb_services/config/environment.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from db_config import DBConfig
+import json
+
+class Environment:
+    def __init__(self):
+        self._db_config = DBConfig()
```
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/config/environment.py]: This should be a dependency to test it properly

File: server/lib/python/cartodb_services/cartodb_services/metrics/user.py [hunk 3]
```diff
@@ -10,9 +11,9 @@ class UserMetricsService:
     SERVICE_HERE_ISOLINES = 'here_isolines'
     DAY_OF_MONTH_ZERO_PADDED = '%d'
 
-    def __init__(self, user_geocoder_config, redis_connection):
+    def __init__(self, user_geocoder_config):
         self._user_geocoder_config = user_geocoder_config
-        self._redis_connection = redis_connection
+        self._redis_connection = RedisConnectionFactory.get_metrics_connection(user_geocoder_config.username)
```
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/user.py]: I assume this is a first iteration, and this is going to be deleted in next iterations

File: server/extension/sql/20_geocode_street.sql [hunk 4]
```diff
@@ -141,14 +142,15 @@ RETURNS Geometry AS $$
   from cartodb_services.mapzen.types import country_to_iso3
   from cartodb_services.metrics import QuotaService
   from cartodb_services.tools import Logger,LoggerConfig
-
-  redis_conn = GD["redis_connection_{0}".format(username)]['redis_metrics_connection']
-  user_geocoder_config = GD["user_geocoder_config_{0}".format(username)]
+  from cartodb_services.tools.redis_tools import RedisConnectionFactory
```
Comment (ethervoid) [inline server/extension/sql/20_geocode_street.sql]: Why is this here? is not used

Comment (iriberri) [inline server/extension/sql/20_geocode_street.sql]: Yep. You're right. This comes from a previous interation where the Redis connection was being instantiated to be sent to QuotaService, but it should take care of getting the connections by itself.

Comment (rafatower) [inline server/extension/sql/20_geocode_street.sql]: you're right, this comes from a previous step in the refactor. Thanks!

File: server/lib/python/cartodb_services/cartodb_services/config/environment.py (hunk 2 above)
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/config/environment.py]: alright

File: server/lib/python/cartodb_services/cartodb_services/metrics/user.py (hunk 3 above)
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/user.py]: most likely not deleted but modified: abstracted and hidden behind a factory that will return one implementation or another depending on config

as stated in the review comment: this is by no means final, just wanted you to :eyes: into it before going any further as things will be much harder to review in later phases

File: server/lib/python/cartodb_services/cartodb_services/tools/redis_tools.py [hunk 5]
```diff
@@ -1,8 +1,31 @@
...
 import json
+import plpy
 
 
+class RedisConnectionFactory:
+
+    @classmethod
+    def get_metadata_connection(cls, username):
+        cache_key = "redis_connection_{0}_metadata".format(username)
+        if cache_key in plpy.GD:
```
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/tools/redis_tools.py]: What I mean, is that with the current version we cache the connection and the config (for example GeocoderConfig) because config doesn't change so we avoid to query Redis every time to get the config.

Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/tools/redis_tools.py]: Point taken. That said I will not mind about (early) optimizations until they prove to be a problem. Anyway caching is an implementation detail that will be isolated rather than scattered in `GD` usage

**Comment by rafatower** [comment]
@ethervoid & @jgoizueta the code WIP in case you want to take a look. I'll shuffle things around for sure, particularly object instantiation

File: server/extension/sql/20_geocode_street.sql [hunk 6]
```diff
@@ -165,16 +165,16 @@ RETURNS Geometry AS $$
   configs = request.configs
 
   # TODO change to hires_geocoder_config = MapzenGeocoderConfigFactory.get(request)
-  user_geocoder_config = MapzenGeocoderConfigFactory(configs).get(user)
+  mapzen_geocoder_config = MapzenGeocoderConfigFactory(configs).get(user)
 
-
-  quota_service = QuotaService(user_geocoder_config)
+  from cartodb_services.refactor.quota.mapzen_quota_service import MapzenGeocoderQuotaService
```
Comment (iriberri) [inline server/extension/sql/20_geocode_street.sql]: I'm missing the actual module restructuration with this commit, but I guess it's coming! ✨

Comment (rafatower) [inline server/extension/sql/20_geocode_street.sql]: Actually I prefer to work that into another branch and maybe close this pull request. I'll brief you on the details later on.

**Comment by iriberri** [comment]
I'm closing this PR, we're moving to branch `redis-refactor-take2`. Please don't delete this branch!
